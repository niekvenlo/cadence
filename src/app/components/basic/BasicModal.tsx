"use client";
import styled from "styled-components";
import { useRef, useEffect } from "react";

const CloseButton = styled.button`
  position: absolute;
  translate: 3em -3em;
  padding: 0.5em;
  width: 3em;
  height: 3em;
  background: unset;
  border: none;
  cursor: pointer;
`;

export default function BasicModal({
  children,
  isOpen,
  closeOnBackdropClick,
  requestClose,
}: {
  children: React.ReactNode;
  isOpen?: boolean;
  closeOnBackdropClick?: boolean;
  requestClose: () => void;
}) {
  const dialogRef = useRef(null);
  useEffect(() => {
    const dialog = dialogRef.current as unknown as HTMLDialogElement;
    if (!dialog) return;
    if (isOpen) {
      dialog.showModal();
    } else {
      dialog.close();
    }
  }, [isOpen]);
  return (
    <dialog
      ref={dialogRef}
      onClick={({ target }) => {
        if (!closeOnBackdropClick) return;
        if (!(target instanceof HTMLElement)) return;
        if (target?.nodeName === "DIALOG") {
          requestClose();
        }
      }}
    >
      <CloseButton
        onClick={() => requestClose()}
        aria-label="Close"
        aria-hidden="true"
      >
        <svg viewBox="0 0 44 44" aria-hidden="true" focusable="false">
          <path d="M0.549989 4.44999L4.44999 0.549988L43.45 39.55L39.55 43.45L0.549989 4.44999Z" />
          <path d="M39.55 0.549988L43.45 4.44999L4.44999 43.45L0.549988 39.55L39.55 0.549988Z" />
        </svg>
      </CloseButton>
      <section>{children}</section>
    </dialog>
  );
}
