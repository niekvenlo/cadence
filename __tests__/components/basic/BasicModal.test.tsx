import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import BasicModal from "../../../src/app/components/basic/BasicModal";

describe("BasicModal", () => {
  beforeEach(() => {
    HTMLDialogElement.prototype.showModal = jest.fn();
    HTMLDialogElement.prototype.close = jest.fn();
  });
  it("renders an open modal", () => {
    render(
      <BasicModal isOpen requestClose={() => {}}>
        contents
      </BasicModal>
    );
    const dialog = screen.getByRole("dialog", { hidden: true });

    expect(dialog).toBeInTheDocument();
    expect(HTMLDialogElement.prototype.showModal).toHaveBeenCalled();
    expect(HTMLDialogElement.prototype.close).not.toHaveBeenCalled();
    expect(dialog).toMatchSnapshot();
  });
  it("renders a closed modal", () => {
    render(<BasicModal requestClose={() => {}}>contents</BasicModal>);
    const dialog = screen.getByRole("dialog", { hidden: true });

    expect(dialog).toBeInTheDocument();
    expect(HTMLDialogElement.prototype.close).toHaveBeenCalled();
    expect(HTMLDialogElement.prototype.showModal).not.toHaveBeenCalled();
    expect(dialog).toMatchSnapshot();
  });
  it("requests a close when the close button is clicked", () => {
    const requestClose = jest.fn();
    render(<BasicModal requestClose={requestClose}>contents</BasicModal>);
    const closeButton = screen.getByLabelText("Close");
    closeButton.click();

    expect(closeButton).toBeInTheDocument();
    expect(requestClose).toHaveBeenCalled();
  });
});
