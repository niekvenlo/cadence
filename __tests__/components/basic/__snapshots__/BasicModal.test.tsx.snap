// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicModal renders a closed modal 1`] = `
<dialog>
  <button
    aria-hidden="true"
    aria-label="Close"
    class="sc-17d31410-0 jtpurn"
  >
    <svg
      aria-hidden="true"
      focusable="false"
      viewBox="0 0 44 44"
    >
      <path
        d="M0.549989 4.44999L4.44999 0.549988L43.45 39.55L39.55 43.45L0.549989 4.44999Z"
      />
      <path
        d="M39.55 0.549988L43.45 4.44999L4.44999 43.45L0.549988 39.55L39.55 0.549988Z"
      />
    </svg>
  </button>
  <section>
    contents
  </section>
</dialog>
`;

exports[`BasicModal renders an open modal 1`] = `
<dialog>
  <button
    aria-hidden="true"
    aria-label="Close"
    class="sc-17d31410-0 jtpurn"
  >
    <svg
      aria-hidden="true"
      focusable="false"
      viewBox="0 0 44 44"
    >
      <path
        d="M0.549989 4.44999L4.44999 0.549988L43.45 39.55L39.55 43.45L0.549989 4.44999Z"
      />
      <path
        d="M39.55 0.549988L43.45 4.44999L4.44999 43.45L0.549988 39.55L39.55 0.549988Z"
      />
    </svg>
  </button>
  <section>
    contents
  </section>
</dialog>
`;
